service/health/types.go

```diff
@@ -1,32 +1,55 @@
 package health
 
 import (
+	"strings"
+
 	v1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/resource"
 
 	"github.com/giantswarm/health-service/service/health/key"
 )
 
-func NewClusterStatus(cluster v1alpha1.StatusCluster) ClusterStatus {
+func NewClusterStatus(cluster v1alpha1.StatusCluster, nodes []v1.Node) ClusterStatus {
 	health := key.Default
-
-	desiredNodes := cluster.Scaling.DesiredCapacity
-	currentNodes := len(cluster.Nodes)
-
-	if currentNodes < desiredNodes {
-		health = key.Yellow
-	}
-
+	roleHealthCounts := map[string]map[key.Health]int{}
+	workerCount := 0
 	creating := cluster.HasCreatingCondition()
 	updating := cluster.HasUpdatingCondition()
-	deleted := cluster.HasDeletedCondition()
 	deleting := cluster.HasDeletingCondition()
 
-	if deleted || deleting {
-		health = key.Red
-	} else if updating || creating { // TODO: Account for draining/scaling
-		health = key.Yellow
+	for _, node := range nodes {
+		role := nodeRole(node.Labels)
+		nodeHealth := calculateNodeHealth(node)
+		if roleHealthCounts[role] == nil {
+			roleHealthCounts[role] = map[key.Health]int{}
+		}
+		roleHealthCounts[role][nodeHealth]++
+		if role == "worker" {
+			workerCount++
+		} else if role == "master" && nodeHealth != key.Green {
+			if updating {
+				health = key.Yellow
+			} else {
+				health = key.Red
+			}
+		}
+	}
+
+	if health != key.Red {
+		if creating || deleting {
+			health = key.Yellow
+		}
+		healthyWorkersProportion := float32(roleHealthCounts["worker"][key.Green]) / float32(workerCount)
+		if healthyWorkersProportion < 0.75 {
+			health = key.Yellow
+		}
+		if workerCount >= 20 && healthyWorkersProportion < 0.5 {
+			health = key.Red
+		}
+		if roleHealthCounts["worker"][key.Green] < 3 {
```

Comment:
No objection here because I know from the spec doc that it was decided any cluster with < 3 workers is unhealthy -- just wondering is that for a particular reason or just best practice?